{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 366,
   "id": "3a6d4798-9cd6-407f-bc8d-85ec7bd3e132",
   "metadata": {},
   "outputs": [],
   "source": [
    "import biom\n",
    "from biom import Table\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from sklearn.model_selection import KFold\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "from sklearn.metrics import accuracy_score, roc_auc_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 285,
   "id": "b2a31750-d1af-40d7-9a69-01b4fb0afcd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.read_excel(\"../../data/metadata.xls\", index_col=0)\n",
    "OTUs_table = biom.load_table(\"../../data/projects/table_6721_2378.biom\")\n",
    "metadata = metadata.loc[OTUs_table.ids(axis=\"sample\")]\n",
    "metadata = metadata.loc[metadata.study.values != \"PRJEB2165\"]\n",
    "metadata = metadata.loc[metadata.study.values != \"PRJNA385004\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "id": "7a6c3cc8-71ac-4b8f-ae61-2a7472d05c5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# metadata = metadata.loc[metadata.fber_type != \"starch-entrapped-microspheres\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 287,
   "id": "01583d76-39f1-4e19-b29f-ff7fa8065f0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "study_list = metadata.study.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "id": "1551adc7-ca99-477c-97b0-8c4a76fa9bff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['PRJEB41443', 'PRJNA560950', 'PRJNA780023', 'PRJNA891951',\n",
       "       'PRJNA293971', 'PRJNA306884', 'PRJNA428736'], dtype=object)"
      ]
     },
     "execution_count": 288,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "study_list"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d8af802-f140-4ca0-9891-af5b7628aca2",
   "metadata": {},
   "source": [
    "# Species"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36a6b515-7eaf-4e7c-8d15-f2c978364be0",
   "metadata": {},
   "source": [
    "### Within-study cross validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "id": "75be6a5d-6069-4ff0-b3d9-aaffc3ea27b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "rf_model = RandomForestClassifier(n_estimators=500, random_state=42, class_weight='balanced')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "id": "2ccaef74-2bbe-4eb2-882c-4f29806633f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "kf = KFold(n_splits=5, shuffle=True, random_state=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "id": "eb9293f2-eaef-4c57-a9c4-9518a3fe7b6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "study_names = []\n",
    "fold_list = []\n",
    "auc_list = []\n",
    "for i in study_list:\n",
    "    metadata_pick = metadata.loc[metadata.study == i]\n",
    "    inter_sid = np.intersect1d(OTUs_table.ids(axis=\"sample\"), metadata_pick.index.values)\n",
    "    OTUs_table_pick = OTUs_table.filter(inter_sid, axis=\"sample\", inplace=False)\n",
    "    OTUs_table_pick = OTUs_table_pick.norm(axis='sample', inplace=False)\n",
    "    OTUs_table_pick = OTUs_table_pick.to_dataframe().T\n",
    "    \n",
    "    host_id = metadata_pick.subject_id.unique()\n",
    "    for fold, (train_idx, test_idx) in enumerate(kf.split(host_id)):\n",
    "        train_sid = metadata_pick.loc[[i in host_id[train_idx] for i in metadata_pick.subject_id.values]].index.values\n",
    "        test_sid = metadata_pick.loc[[i not in host_id[train_idx] for i in metadata_pick.subject_id.values]].index.values\n",
    "        \n",
    "        X_train = OTUs_table_pick.loc[train_sid]\n",
    "        X_test = OTUs_table_pick.loc[test_sid]\n",
    "        y_train = metadata_pick.loc[train_sid].group.values\n",
    "        y_test = metadata_pick.loc[test_sid].group.values\n",
    "        \n",
    "        rf_model.fit(X_train, y_train)\n",
    "        y_pred = rf_model.predict(X_test)\n",
    "        y_prob = rf_model.predict_proba(X_test)\n",
    "    \n",
    "        auc = roc_auc_score(y_test, y_prob[:, 1])\n",
    "        \n",
    "        study_names.append(i)\n",
    "        fold_list.append(f\"fold_{fold+1}\")\n",
    "        auc_list.append(auc)\n",
    "        \n",
    "results_cv = pd.DataFrame({\"study\": study_names, \"fold\": fold_list, \"auc\": auc_list})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "id": "7c0e08e3-1d8e-461d-bb96-3db4328a0ccc",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_cv.to_csv(\"../../data/classification/results_cv.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fcda9e4-1cab-4820-99b2-b540e1b321e5",
   "metadata": {},
   "source": [
    "### Cross-study validation (CSV)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "id": "4d8fdeb8-2649-41e4-a61a-0acd54436d29",
   "metadata": {},
   "outputs": [],
   "source": [
    "rf_model = RandomForestClassifier(n_estimators=500, random_state=42, class_weight='balanced')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "id": "5f8cc752-01a1-43f8-bc68-87d19f3e9d1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_study = []\n",
    "test_study = []\n",
    "auc_list = []\n",
    "for i in study_list:\n",
    "    train_sid = metadata.loc[metadata.study.values == i].index.values\n",
    "    for j in study_list:\n",
    "        if i != j:\n",
    "            test_sid = metadata.loc[metadata.study.values == j].index.values\n",
    "            \n",
    "            X_train = OTUs_table.filter(train_sid, axis=\"sample\", inplace=False)\n",
    "            X_train.norm(axis='sample')\n",
    "            X_train = X_train.to_dataframe().T\n",
    "            \n",
    "            X_test = OTUs_table.filter(test_sid, axis=\"sample\", inplace=False)\n",
    "            X_test.norm(axis='sample')\n",
    "            X_test = X_test.to_dataframe().T\n",
    "            \n",
    "            y_train = metadata.loc[train_sid].group.values\n",
    "            y_test = metadata.loc[test_sid].group.values\n",
    "            \n",
    "            rf_model.fit(X_train, y_train)\n",
    "            y_pred = rf_model.predict(X_test)\n",
    "            y_prob = rf_model.predict_proba(X_test)\n",
    "        \n",
    "            auc = roc_auc_score(y_test, y_prob[:, 1])\n",
    "            \n",
    "            train_study.append(i)\n",
    "            test_study.append(j)\n",
    "            auc_list.append(auc)\n",
    "\n",
    "results_csv = pd.DataFrame({\"train_study\": train_study, \"test_study\": test_study, \"auc\": auc_list})\n",
    "results_csv.to_csv(\"../../data/classification/results_csv.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29a83304-a202-4c40-acec-ec9d88536c5c",
   "metadata": {},
   "source": [
    "### Leave-one-study-out validation (LOSO)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "id": "a8dfa4ca-1ce4-4bd9-9cb9-1a5932a1247e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "test_study = []\n",
    "auc_list = []\n",
    "for i in study_list:\n",
    "    df = metadata.loc[metadata.study.values != i]\n",
    "    # train_sid_1 = df.loc[df.group.values == 0].drop_duplicates(subset=['subject_id']).index.values\n",
    "    # train_sid_2 = df.loc[df.group.values == 1].drop_duplicates(subset=['subject_id']).index.values\n",
    "    # train_sid = list(train_sid_1) + list(train_sid_2)\n",
    "    train_sid = metadata.loc[metadata.study.values != i].index.values\n",
    "    test_sid = metadata.loc[metadata.study.values == i].index.values\n",
    "    \n",
    "    X_train = OTUs_table.filter(train_sid, axis=\"sample\", inplace=False)\n",
    "    X_train.remove_empty()\n",
    "    fid = X_train.ids(axis=\"observation\")\n",
    "    X_train.norm(axis='sample')\n",
    "    X_train = X_train.to_dataframe().T\n",
    "    \n",
    "    X_test = OTUs_table.filter(test_sid, axis=\"sample\", inplace=False)\n",
    "    X_test = X_test.filter(fid, axis=\"observation\", inplace=False)\n",
    "    X_test.norm(axis='sample')\n",
    "    X_test = X_test.to_dataframe().T\n",
    "    \n",
    "    y_train = metadata.loc[train_sid].group.values\n",
    "    y_test = metadata.loc[test_sid].group.values\n",
    "    \n",
    "    rf_model.fit(X_train, y_train)\n",
    "    y_pred = rf_model.predict(X_test)\n",
    "    y_prob = rf_model.predict_proba(X_test)\n",
    "\n",
    "    auc = roc_auc_score(y_test, y_prob[:, 1])\n",
    "\n",
    "    test_study.append(i)\n",
    "    auc_list.append(auc)\n",
    "\n",
    "results_loso = pd.DataFrame({\"test_study\": test_study, \"auc\": auc_list})\n",
    "results_loso.to_csv(\"../../data/classification/results_loso.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "id": "9b364fb1-613a-4e60-8308-74b49301eab5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>test_study</th>\n",
       "      <th>auc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJEB41443</td>\n",
       "      <td>0.750678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRJNA560950</td>\n",
       "      <td>0.701942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PRJNA780023</td>\n",
       "      <td>0.508412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PRJNA891951</td>\n",
       "      <td>0.559547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PRJNA293971</td>\n",
       "      <td>0.731875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PRJNA306884</td>\n",
       "      <td>0.825470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PRJNA428736</td>\n",
       "      <td>0.652200</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    test_study       auc\n",
       "0   PRJEB41443  0.750678\n",
       "1  PRJNA560950  0.701942\n",
       "2  PRJNA780023  0.508412\n",
       "3  PRJNA891951  0.559547\n",
       "4  PRJNA293971  0.731875\n",
       "5  PRJNA306884  0.825470\n",
       "6  PRJNA428736  0.652200"
      ]
     },
     "execution_count": 277,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_loso"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73a8e06a-f0fb-41a5-8e6f-a0f484d23df7",
   "metadata": {},
   "source": [
    "### Biomark"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 315,
   "id": "8b54a2b0-2218-4f33-8675-4293cadf3786",
   "metadata": {},
   "outputs": [],
   "source": [
    "netmoss = pd.read_csv(\"/home/dongbiao/word_embedding_microbiome/programe_test/dietary_fber/data/network/netmoss.csv\")\n",
    "netmoss = netmoss.sort_values(\"NetMoss_Score\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 318,
   "id": "db4bc722-60c0-42a7-90c1-b39161614aeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# lind_res = {}\n",
    "# for i in study_list:\n",
    "#     lind_res[i] = pd.read_csv(f\"../../data/difference_analysis/linda_single_study_{i}.csv\")\n",
    "\n",
    "# lind_res = pd.concat(lind_res, axis=0, ignore_index=True)\n",
    "# lind_res = lind_res.loc[lind_res.padj < 0.05]\n",
    "# # lind_res = lind_res.loc[lind_res.reject == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 316,
   "id": "39a1f578-3542-4fe0-9d2f-d827aa767c2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# FDR = [0.05, 0.1, 0.2, 0.4]\n",
    "# for j in FDR:\n",
    "# combine = pd.read_csv(\"../../data/difference_analysis/combine_results.csv\", index_col=0)\n",
    "# combine = combine.loc[combine.padj.values < j]\n",
    "# linda = pd.read_csv(\"../../data/difference_analysis/combine_results.csv\", index_col=0)\n",
    "# linda = linda.loc[linda.padj.values < j]\n",
    "\n",
    "test_study = []\n",
    "auc_list = []\n",
    "# fid = lind_res[\"Unnamed: 0\"].values\n",
    "fid = netmoss.taxon_names.values\n",
    "OTUs_table_biomark =  OTUs_table.filter(fid, axis=\"observation\", inplace=False)\n",
    "# OTUs_table_biomark.remove_empty()\n",
    "\n",
    "for i in study_list:\n",
    "    df = metadata.loc[metadata.study.values != i]\n",
    "    train_sid = metadata.loc[metadata.study.values != i].index.values\n",
    "    test_sid = metadata.loc[metadata.study.values == i].index.values\n",
    "    \n",
    "    X_train = OTUs_table_biomark.filter(train_sid, axis=\"sample\", inplace=False)\n",
    "    X_train.remove_empty()\n",
    "    fid = X_train.ids(axis=\"observation\")\n",
    "    X_train.norm(axis='sample')\n",
    "    X_train = X_train.to_dataframe().T\n",
    "    \n",
    "    X_test = OTUs_table_biomark.filter(test_sid, axis=\"sample\", inplace=False)\n",
    "    X_test = X_test.filter(fid, axis=\"observation\", inplace=False)\n",
    "    X_test.norm(axis='sample')\n",
    "    X_test = X_test.to_dataframe().T\n",
    "    \n",
    "    y_train = metadata.loc[X_train.index.values].group.values\n",
    "    y_test = metadata.loc[X_test.index.values].group.values\n",
    "    \n",
    "    rf_model.fit(X_train, y_train)\n",
    "    y_pred = rf_model.predict(X_test)\n",
    "    y_prob = rf_model.predict_proba(X_test)\n",
    "\n",
    "    auc = roc_auc_score(y_test, y_prob[:, 1])\n",
    "\n",
    "    test_study.append(i)\n",
    "    auc_list.append(auc)\n",
    "\n",
    "results_biomark = pd.DataFrame({\"test_study\": test_study, \"auc\": auc_list})\n",
    "results_biomark.to_csv(f\"../../data/classification/results_loso_biomark.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 317,
   "id": "ae2ed20e-e159-4f73-9ac6-f70b1aac1b02",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>test_study</th>\n",
       "      <th>auc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJEB41443</td>\n",
       "      <td>0.730531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRJNA560950</td>\n",
       "      <td>0.707466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PRJNA780023</td>\n",
       "      <td>0.502264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PRJNA891951</td>\n",
       "      <td>0.575638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PRJNA293971</td>\n",
       "      <td>0.751250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PRJNA306884</td>\n",
       "      <td>0.829487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PRJNA428736</td>\n",
       "      <td>0.678389</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    test_study       auc\n",
       "0   PRJEB41443  0.730531\n",
       "1  PRJNA560950  0.707466\n",
       "2  PRJNA780023  0.502264\n",
       "3  PRJNA891951  0.575638\n",
       "4  PRJNA293971  0.751250\n",
       "5  PRJNA306884  0.829487\n",
       "6  PRJNA428736  0.678389"
      ]
     },
     "execution_count": 317,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_biomark"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c433e385-4484-4d2c-b490-830624eb5258",
   "metadata": {},
   "source": [
    "# CAZy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "id": "75f4d228-a240-4bdc-8424-7d504cb91599",
   "metadata": {},
   "outputs": [],
   "source": [
    "rf_model = RandomForestClassifier(n_estimators=500, random_state=42, class_weight='balanced')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "id": "29472473-c9f2-410f-ad6b-c58488040750",
   "metadata": {},
   "outputs": [],
   "source": [
    "cazy_table = pd.read_csv(\"../../data/picrust2/CAZY_metagenome_out/pred_metagenome_unstrat.tsv\", sep=\"\\t\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "id": "0778dc7c-5b76-409d-b9d3-9b4d2cb9a113",
   "metadata": {},
   "outputs": [],
   "source": [
    "cazy_table = cazy_table.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "id": "7c991a27-b551-4d9f-a27c-9d4aa06dae4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cazy_table = cazy_table.apply(lambda x: x / x.sum(), axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99c8706b-4d31-4621-880e-286f3edf1681",
   "metadata": {},
   "source": [
    "### Within-study cross validationÂ¶"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "id": "a750c72a-8828-4fc2-9f83-ccd5eb8e1857",
   "metadata": {},
   "outputs": [],
   "source": [
    "study_names = []\n",
    "fold_list = []\n",
    "auc_list = []\n",
    "for i in study_list:\n",
    "    metadata_pick = metadata.loc[metadata.study == i]\n",
    "    inter_sid = np.intersect1d(cazy_table.index.values, metadata_pick.index.values)\n",
    "    cazy_table_pick = cazy_table.loc[inter_sid]\n",
    "    \n",
    "    host_id = metadata_pick.subject_id.unique()\n",
    "    for fold, (train_idx, test_idx) in enumerate(kf.split(host_id)):\n",
    "        train_sid = metadata_pick.loc[[i in host_id[train_idx] for i in metadata_pick.subject_id.values]].index.values\n",
    "        test_sid = metadata_pick.loc[[i not in host_id[train_idx] for i in metadata_pick.subject_id.values]].index.values\n",
    "        \n",
    "        X_train = cazy_table_pick.loc[train_sid]\n",
    "        X_test = cazy_table_pick.loc[test_sid]\n",
    "        y_train = metadata_pick.loc[train_sid].group.values\n",
    "        y_test = metadata_pick.loc[test_sid].group.values\n",
    "        \n",
    "        rf_model.fit(X_train, y_train)\n",
    "        y_pred = rf_model.predict(X_test)\n",
    "        y_prob = rf_model.predict_proba(X_test)\n",
    "    \n",
    "        auc = roc_auc_score(y_test, y_prob[:, 1])\n",
    "        \n",
    "        study_names.append(i)\n",
    "        fold_list.append(f\"fold_{fold+1}\")\n",
    "        auc_list.append(auc)\n",
    "        \n",
    "results_cv = pd.DataFrame({\"study\": study_names, \"fold\": fold_list, \"auc\": auc_list})\n",
    "results_cv.to_csv(\"../../data/classification/results_cazy_cv.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67c3a5f2-7705-415f-9241-0778a33bfb69",
   "metadata": {},
   "source": [
    "### Cross-study validation (CSV)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "id": "24fd773d-168e-408b-b6fd-ba2fcde759ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_study = []\n",
    "test_study = []\n",
    "auc_list = []\n",
    "for i in study_list:\n",
    "    train_sid = metadata.loc[metadata.study.values == i].index.values\n",
    "    for j in study_list:\n",
    "        if i != j:\n",
    "            test_sid = metadata.loc[metadata.study.values == j].index.values\n",
    "            \n",
    "            X_train = cazy_table.loc[train_sid]\n",
    "            X_test = cazy_table.loc[test_sid]\n",
    "            \n",
    "            y_train = metadata.loc[train_sid].group.values\n",
    "            y_test = metadata.loc[test_sid].group.values\n",
    "            \n",
    "            rf_model.fit(X_train, y_train)\n",
    "            y_pred = rf_model.predict(X_test)\n",
    "            y_prob = rf_model.predict_proba(X_test)\n",
    "        \n",
    "            auc = roc_auc_score(y_test, y_prob[:, 1])\n",
    "            \n",
    "            train_study.append(i)\n",
    "            test_study.append(j)\n",
    "            auc_list.append(auc)\n",
    "\n",
    "results_csv = pd.DataFrame({\"train_study\": train_study, \"test_study\": test_study, \"auc\": auc_list})\n",
    "results_csv.to_csv(\"../../data/classification/results_cazy_csv.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b6cb8b6-7a28-41d3-b81a-ec2f283aec7b",
   "metadata": {},
   "source": [
    "### Leave-one-study-out validation (LOSO)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "id": "09bb5481-a016-44f7-8a0d-3330704c84e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_study = []\n",
    "auc_list = []\n",
    "for i in study_list:\n",
    "    df = metadata.loc[metadata.study.values != i]\n",
    "    train_sid = metadata.loc[metadata.study.values != i].index.values\n",
    "    test_sid = metadata.loc[metadata.study.values == i].index.values\n",
    "    \n",
    "    X_train = cazy_table.loc[train_sid]\n",
    "    X_test = cazy_table.loc[test_sid]\n",
    "    \n",
    "    y_train = metadata.loc[train_sid].group.values\n",
    "    y_test = metadata.loc[test_sid].group.values\n",
    "    \n",
    "    rf_model.fit(X_train, y_train)\n",
    "    y_pred = rf_model.predict(X_test)\n",
    "    y_prob = rf_model.predict_proba(X_test)\n",
    "\n",
    "    auc = roc_auc_score(y_test, y_prob[:, 1])\n",
    "\n",
    "    test_study.append(i)\n",
    "    auc_list.append(auc)\n",
    "\n",
    "results_loso = pd.DataFrame({\"test_study\": test_study, \"auc\": auc_list})\n",
    "results_loso.to_csv(\"../../data/classification/results_cazy_loso.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b94cad78-7852-4c1b-88b8-390d7debdba2",
   "metadata": {},
   "source": [
    "### Biomarker"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 323,
   "id": "b837a6dd-5d98-4d75-9683-ce4632f94df0",
   "metadata": {},
   "outputs": [],
   "source": [
    "lind_res = {}\n",
    "for i in study_list:\n",
    "    lind_res[i] = pd.read_csv(f\"../../data/difference_analysis/linda_single_study_cazy_{i}.csv\")\n",
    "\n",
    "lind_res = pd.concat(lind_res, axis=0, ignore_index=True)\n",
    "lind_res = lind_res.loc[lind_res.padj < 0.1]\n",
    "    \n",
    "test_study = []\n",
    "auc_list = []\n",
    "fid = lind_res[\"Unnamed: 0\"].values\n",
    "cazy_table_biomark =  cazy_table.loc[:, fid]\n",
    "\n",
    "for i in study_list:\n",
    "    df = metadata.loc[metadata.study.values != i]\n",
    "    train_sid = metadata.loc[metadata.study.values != i].index.values\n",
    "    test_sid = metadata.loc[metadata.study.values == i].index.values\n",
    "    \n",
    "    X_train = cazy_table_biomark.loc[train_sid]\n",
    "    X_test = cazy_table_biomark.loc[test_sid]\n",
    "    \n",
    "    y_train = metadata.loc[X_train.index.values].group.values\n",
    "    y_test = metadata.loc[X_test.index.values].group.values\n",
    "    \n",
    "    rf_model.fit(X_train, y_train)\n",
    "    y_pred = rf_model.predict(X_test)\n",
    "    y_prob = rf_model.predict_proba(X_test)\n",
    "\n",
    "    auc = roc_auc_score(y_test, y_prob[:, 1])\n",
    "\n",
    "    test_study.append(i)\n",
    "    auc_list.append(auc)\n",
    "\n",
    "results_biomark = pd.DataFrame({\"test_study\": test_study, \"auc\": auc_list})\n",
    "results_biomark.to_csv(f\"../../data/classification/results_cazy_loso_biomark.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "832e9526-01f1-4f35-9c15-b4da881dea57",
   "metadata": {},
   "source": [
    "# KO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "id": "4f656a83-926c-44e8-bd57-3543c8eaa1e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "ko_table = pd.read_csv(\"../../data/picrust2/KO_metagenome_out/pred_metagenome_unstrat.tsv\", sep=\"\\t\", index_col=0)\n",
    "ko_table = ko_table.T\n",
    "ko_table = ko_table.apply(lambda x: x / x.sum(), axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ba81840-a895-4b6e-b52f-082234d364d5",
   "metadata": {},
   "source": [
    "### Within-study cross validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "id": "9d17d1bc-726e-4bef-a026-a3653f499b43",
   "metadata": {},
   "outputs": [],
   "source": [
    "study_names = []\n",
    "fold_list = []\n",
    "auc_list = []\n",
    "for i in study_list:\n",
    "    metadata_pick = metadata.loc[metadata.study == i]\n",
    "    inter_sid = np.intersect1d(ko_table.index.values, metadata_pick.index.values)\n",
    "    ko_table_pick = ko_table.loc[inter_sid]\n",
    "    \n",
    "    host_id = metadata_pick.subject_id.unique()\n",
    "    for fold, (train_idx, test_idx) in enumerate(kf.split(host_id)):\n",
    "        train_sid = metadata_pick.loc[[i in host_id[train_idx] for i in metadata_pick.subject_id.values]].index.values\n",
    "        test_sid = metadata_pick.loc[[i not in host_id[train_idx] for i in metadata_pick.subject_id.values]].index.values\n",
    "        \n",
    "        X_train = ko_table_pick.loc[train_sid]\n",
    "        X_test = ko_table_pick.loc[test_sid]\n",
    "        y_train = metadata_pick.loc[train_sid].group.values\n",
    "        y_test = metadata_pick.loc[test_sid].group.values\n",
    "        \n",
    "        rf_model.fit(X_train, y_train)\n",
    "        y_pred = rf_model.predict(X_test)\n",
    "        y_prob = rf_model.predict_proba(X_test)\n",
    "    \n",
    "        auc = roc_auc_score(y_test, y_prob[:, 1])\n",
    "        \n",
    "        study_names.append(i)\n",
    "        fold_list.append(f\"fold_{fold+1}\")\n",
    "        auc_list.append(auc)\n",
    "        \n",
    "results_cv = pd.DataFrame({\"study\": study_names, \"fold\": fold_list, \"auc\": auc_list})\n",
    "results_cv.to_csv(\"../../data/classification/results_ko_cv.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9152574d-18a8-497d-ac3e-e886b6364ef2",
   "metadata": {},
   "source": [
    "### Cross-study validation (CSV)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "id": "bbd8df1b-a3fc-4622-9e67-0474b55b0d7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_study = []\n",
    "test_study = []\n",
    "auc_list = []\n",
    "for i in study_list:\n",
    "    train_sid = metadata.loc[metadata.study.values == i].index.values\n",
    "    for j in study_list:\n",
    "        if i != j:\n",
    "            test_sid = metadata.loc[metadata.study.values == j].index.values\n",
    "            \n",
    "            X_train = ko_table.loc[train_sid]\n",
    "            X_test = ko_table.loc[test_sid]\n",
    "            \n",
    "            y_train = metadata.loc[train_sid].group.values\n",
    "            y_test = metadata.loc[test_sid].group.values\n",
    "            \n",
    "            rf_model.fit(X_train, y_train)\n",
    "            y_pred = rf_model.predict(X_test)\n",
    "            y_prob = rf_model.predict_proba(X_test)\n",
    "        \n",
    "            auc = roc_auc_score(y_test, y_prob[:, 1])\n",
    "            \n",
    "            train_study.append(i)\n",
    "            test_study.append(j)\n",
    "            auc_list.append(auc)\n",
    "\n",
    "results_csv = pd.DataFrame({\"train_study\": train_study, \"test_study\": test_study, \"auc\": auc_list})\n",
    "results_csv.to_csv(\"../../data/classification/results_ko_csv.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03b09b95-3e9f-41db-b1a2-0bfee27caea7",
   "metadata": {},
   "source": [
    "### Leave-one-study-out validation (LOSO)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "id": "9e8dfe77-c2e0-4a67-80f6-7724bf420ea5",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_study = []\n",
    "auc_list = []\n",
    "for i in study_list:\n",
    "    df = metadata.loc[metadata.study.values != i]\n",
    "    train_sid = metadata.loc[metadata.study.values != i].index.values\n",
    "    test_sid = metadata.loc[metadata.study.values == i].index.values\n",
    "    test_sid = metadata.loc[metadata.study.values == i].index.values\n",
    "    \n",
    "    X_train = ko_table.loc[train_sid]\n",
    "    X_test = ko_table.loc[test_sid]\n",
    "    \n",
    "    y_train = metadata.loc[train_sid].group.values\n",
    "    y_test = metadata.loc[test_sid].group.values\n",
    "    \n",
    "    rf_model.fit(X_train, y_train)\n",
    "    y_pred = rf_model.predict(X_test)\n",
    "    y_prob = rf_model.predict_proba(X_test)\n",
    "\n",
    "    auc = roc_auc_score(y_test, y_prob[:, 1])\n",
    "\n",
    "    test_study.append(i)\n",
    "    auc_list.append(auc)\n",
    "\n",
    "results_loso = pd.DataFrame({\"test_study\": test_study, \"auc\": auc_list})\n",
    "results_loso.to_csv(\"../../data/classification/results_ko_loso.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2a05205-5f0f-4512-8136-24b806ec4f18",
   "metadata": {},
   "source": [
    "### Biomarker"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 321,
   "id": "db7ec0f7-3974-4fc8-a6c4-bc328bc9a658",
   "metadata": {},
   "outputs": [],
   "source": [
    "lind_res = {}\n",
    "for i in study_list:\n",
    "    lind_res[i] = pd.read_csv(f\"../../data/difference_analysis/linda_single_study_ko_{i}.csv\")\n",
    "\n",
    "lind_res = pd.concat(lind_res, axis=0, ignore_index=True)\n",
    "lind_res = lind_res.loc[lind_res.padj < 0.05]\n",
    "    \n",
    "test_study = []\n",
    "auc_list = []\n",
    "fid = lind_res[\"Unnamed: 0\"].values\n",
    "ko_table_biomark =  ko_table.loc[:, fid]\n",
    "\n",
    "for i in study_list:\n",
    "    df = metadata.loc[metadata.study.values != i]\n",
    "    train_sid = metadata.loc[metadata.study.values != i].index.values\n",
    "    test_sid = metadata.loc[metadata.study.values == i].index.values\n",
    "    \n",
    "    X_train = ko_table_biomark.loc[train_sid]\n",
    "    X_test = ko_table_biomark.loc[test_sid]\n",
    "    \n",
    "    y_train = metadata.loc[X_train.index.values].group.values\n",
    "    y_test = metadata.loc[X_test.index.values].group.values\n",
    "    \n",
    "    rf_model.fit(X_train, y_train)\n",
    "    y_pred = rf_model.predict(X_test)\n",
    "    y_prob = rf_model.predict_proba(X_test)\n",
    "\n",
    "    auc = roc_auc_score(y_test, y_prob[:, 1])\n",
    "\n",
    "    test_study.append(i)\n",
    "    auc_list.append(auc)\n",
    "\n",
    "results_biomark = pd.DataFrame({\"test_study\": test_study, \"auc\": auc_list})\n",
    "results_biomark.to_csv(f\"../../data/classification/results_ko_loso_biomark.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4792805-8274-48f7-a1ad-f21770716845",
   "metadata": {},
   "source": [
    "### IBD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 389,
   "id": "2fcc6452-9886-4928-adba-f9a998e2906c",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.read_csv(\"/home/dongbiao/word_embedding_microbiome/Classification_prediction/data_cjj/ibd/metadata.tsv\", sep=\"\\t\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 390,
   "id": "1792420e-74bf-42a7-bf6e-890ce4dddd8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "auc_list = []\n",
    "group = []\n",
    "for i in [\"PRJNA324147\", \"PRJNA422193\", \"PRJNA450340\", \"qiita_2538\", \"PRJNA368966\", \"PRJNA431126\", \"qiita_1629\", \"RISK_PRISM_f\"]:\n",
    "    \n",
    "    train = biom.load_table(f\"/home/dongbiao/word_embedding_microbiome/Classification_prediction/data_cjj/ibd/{i}/train.biom\")\n",
    "    train = train.norm(axis='sample', inplace=False)\n",
    "    X_train = train.to_dataframe().T\n",
    "    \n",
    "    test = biom.load_table(f\"/home/dongbiao/word_embedding_microbiome/Classification_prediction/data_cjj/ibd/{i}/test.biom\")\n",
    "    test = test.norm(axis='sample', inplace=False)\n",
    "    X_test = test.to_dataframe().T\n",
    "\n",
    "    y_train = metadata.loc[X_train.index.values].group.values\n",
    "    y_test = metadata.loc[X_test.index.values].group.values\n",
    "    \n",
    "    rf_model.fit(X_train, y_train)\n",
    "    y_pred = rf_model.predict(X_test)\n",
    "    y_prob = rf_model.predict_proba(X_test)\n",
    "\n",
    "    auc = roc_auc_score(y_test, y_prob[:, 1])\n",
    "    \n",
    "    auc_list.append(auc)\n",
    "    group.append(\"RF\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 386,
   "id": "c926d402-f2cf-4b27-b342-e0fa4bfcc298",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for i in [\"PRJNA324147\", \"PRJNA422193\", \"PRJNA450340\", \"qiita_2538\", \"PRJNA368966\", \"PRJNA431126\", \"qiita_1629\", \"RISK_PRISM_f\"]:\n",
    "\n",
    "#     train = biom.load_table(f\"/home/dongbiao/word_embedding_microbiome/Classification_prediction/data_cjj/ibd/{i}/train.biom\")\n",
    "#     X_train = train.to_dataframe().T\n",
    "#     values = X_train.values.flatten()\n",
    "#     np.random.shuffle(values)\n",
    "#     shuffled_values = values.reshape(X_train.shape)\n",
    "#     X_train_shuffled = pd.DataFrame(shuffled_values, index=X_train.index, columns=X_train.columns)\n",
    "#     shuffled_table = biom.Table(data=X_train_shuffled.T.values, \n",
    "#                                 observation_ids=X_train_shuffled.columns.tolist(), \n",
    "#                                 sample_ids=X_train_shuffled.index.tolist())\n",
    "#     with biom.util.biom_open(f\"/home/dongbiao/word_embedding_microbiome/Classification_prediction/data_cjj/ibd/{i}/table_shuffled.biom\", 'w') as f:\n",
    "#         shuffled_table.to_hdf5(f, 'example')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 391,
   "id": "9b7670a7-5612-41b0-83c1-3364c8454171",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in [\"PRJNA324147\", \"PRJNA422193\", \"PRJNA450340\", \"qiita_2538\", \"PRJNA368966\", \"PRJNA431126\", \"qiita_1629\", \"RISK_PRISM_f\"]:\n",
    "    \n",
    "    train = biom.load_table(f\"/home/dongbiao/word_embedding_microbiome/Classification_prediction/data_cjj/ibd/{i}/train.biom\")\n",
    "    train = train.norm(axis='sample', inplace=False)\n",
    "    X_train = train.to_dataframe().T\n",
    "    X_train = X_train.apply(np.random.permutation, axis=1, result_type='broadcast')\n",
    "    \n",
    "    test = biom.load_table(f\"/home/dongbiao/word_embedding_microbiome/Classification_prediction/data_cjj/ibd/{i}/test.biom\")\n",
    "    test = test.norm(axis='sample', inplace=False)\n",
    "    X_test = test.to_dataframe().T\n",
    "\n",
    "    y_train = metadata.loc[X_train.index.values].group.values\n",
    "    y_test = metadata.loc[X_test.index.values].group.values\n",
    "    \n",
    "    rf_model.fit(X_train, y_train)\n",
    "    y_pred = rf_model.predict(X_test)\n",
    "    y_prob = rf_model.predict_proba(X_test)\n",
    "\n",
    "    auc = roc_auc_score(y_test, y_prob[:, 1])\n",
    "    \n",
    "    auc_list.append(auc)\n",
    "    group.append(\"RF_Row-wise_shuffle\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 392,
   "id": "0d62e539-e8c2-4e83-9d1f-7cbb52256b40",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in [\"PRJNA324147\", \"PRJNA422193\", \"PRJNA450340\", \"qiita_2538\", \"PRJNA368966\", \"PRJNA431126\", \"qiita_1629\", \"RISK_PRISM_f\"]:\n",
    "    \n",
    "    train = biom.load_table(f\"/home/dongbiao/word_embedding_microbiome/Classification_prediction/data_cjj/ibd/{i}/train.biom\")\n",
    "    train = train.norm(axis='sample', inplace=False)\n",
    "    X_train = train.to_dataframe().T\n",
    "    values = X_train.values.flatten()\n",
    "    np.random.shuffle(values)\n",
    "    shuffled_values = values.reshape(X_train.shape)\n",
    "    X_train_shuffled = pd.DataFrame(shuffled_values, index=X_train.index, columns=X_train.columns)\n",
    "    \n",
    "    test = biom.load_table(f\"/home/dongbiao/word_embedding_microbiome/Classification_prediction/data_cjj/ibd/{i}/test.biom\")\n",
    "    test = test.norm(axis='sample', inplace=False)\n",
    "    X_test = test.to_dataframe().T\n",
    "\n",
    "    y_train = metadata.loc[X_train.index.values].group.values\n",
    "    y_test = metadata.loc[X_test.index.values].group.values\n",
    "    \n",
    "    rf_model.fit(X_train_shuffled, y_train)\n",
    "    y_pred = rf_model.predict(X_test)\n",
    "    y_prob = rf_model.predict_proba(X_test)\n",
    "\n",
    "    auc = roc_auc_score(y_test, y_prob[:, 1])\n",
    "    \n",
    "    auc_list.append(auc)\n",
    "    group.append(\"RF_Global_shuffle\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 395,
   "id": "22ea5bee-3a3a-439c-919b-a0f552360212",
   "metadata": {},
   "outputs": [],
   "source": [
    "attention_noembedding_auc = [0.764, 0.624, 0.710, 0.728, 0.683, 0.534, 0.729, 0.559]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 396,
   "id": "8ddb503c-0acc-4f5a-b4b7-b984cf4915aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "auc_list = auc_list + attention_noembedding_auc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 397,
   "id": "d4e43933-424e-4796-aeec-5694d5f5a3b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "group = group + [\"Attention_noembedding\"] * 8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 398,
   "id": "e43106c3-cca4-4702-a468-6651436463cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "RF_shuffle = pd.DataFrame({\"AUC\" : auc_list, \"group\": group})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 400,
   "id": "19165007-ad9f-4040-a399-a3b38c867311",
   "metadata": {},
   "outputs": [],
   "source": [
    "RF_shuffle.to_csv(\"/home/dongbiao/word_embedding_microbiome/Classification_prediction/data_cjj/ibd/RF_shuffle.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8ea1bb2-89a7-4ec1-92f5-faddef3f1e77",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
